# Business Analytics - Assignment 2  

**Assignment Points**: 100  
**Submission**: Provide your answers in this notebook and submit it via iLearn

- Where a question requires a written (text) solution provide your answer in Markdown in appropriate cells under each question.
- Comment out your print statements unless you are explicitly asked to use the print() function. 
- 5 marks will be deducted for printed outputs that are not asked for.

### About the Assignment

- Assignment 2 extends Assignment 1 on credit card applications. 


- For this assignment there are two files in the `data` folder `credit_record.csv` and `application_record.csv` where bank clients are related by the `ID` column.
- In `credit_record.csv` we have the following variables

| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| AMT_INCOME_TOTAL   | Annual income  |  |
| NAME_INCOME_TYPE   | Income Source |  |
| NAME_EDUCATION_TYPE   | Level of Education  |  |
| CODE_GENDER   | Applicant's Gender   |  |
| FLAG_OWN_CAR | Car Ownership |  | 
| CNT_CHILDREN | Number of Children | |
| FLAG_OWN_REALTY | Real Estate Ownership | | 
| NAME_FAMILY_STATUS | Relationship Status | | 
| NAME_HOUSING_TYPE | Housing Type | | 
| DAYS_BIRTH | No. of Days | Count backwards from current day (0), -1 means yesterday
| DAYS_EMPLOYED | No. of Days | Count backwards from current day(0). If positive, it means the person is currently unemployed.
| FLAG_MOBIL | Mobile Phone Ownership | | 
| FLAG_WORK_PHONE | Work Phone Ownership | | 
| FLAG_PHONE | Landline Phone Ownership | | 
| FLAG_EMAIL | Landline Phone Ownership | | 
| OCCUPATION_TYPE | Occupation | | 
| CNT_FAM_MEMBERS | Count of Family Members | |



- In `credit_record.csv` we have the following variables


| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number | |
| MONTHS_BALANCE | Number of months in the past from now when STATUS is measured | 0 = current month, -1 = last month, -2 = two months ago, etc.|
| STATUS | Number of days a payment is past due | 0: 1-29 days past due 1: 30-59 days past due 2: 60-89 days overdue 3: 90-119 days overdue 4: 120-149 days overdue 5: Overdue or bad debts, write-offs for more than 150 days C: paid off that month X: No loan for the month |

---
---

### Task 1: Reading, Summarising and Cleaning Data (Total Marks: 30)



**Question 1.** 

1. Import the `application_record.csv` and `credit_record.csv` files from `data` folder into pandas DataFrames named `df_application` and `df_credit`, respectively. (1 mark)

2. How many rows are there in `df_application` and `df_credit`, respectively? Answer using both print() function and in Markdown text. (1 mark)

3. How many unique bank clients are there in `df_application` and `df_credit`? Answer using both print() function and in Markdown text. (1 mark)

4. Add the records from `df_credit` to `df_application` by merging the data from the two DataFrames on the `ID` column, and output the joint data into a new DataFrame named `df`. Hint: Use `merge` function from pandas by setting `how` parameter to `inner` (4 marks) 

5. How many rows and how many unique clients are there in `df`? (1 mark)

6. How are multiple rows for each `ID` in `df` different? Answer in Markdown text. (2 mark) 

(10 marks)


In [27]:
import pandas as pd

import matplotlib.pyplot as plt

import numpy as np

In [28]:
df_application= pd.read_csv('application_record.csv')
df_credit= pd.read_csv('credit_record.csv')

In [29]:
print(df_application)
print(df_credit)

             ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0       5008804           M            Y               Y           0.0   
1       5008805           M            Y               Y           0.0   
2       5008806           M            Y               Y           0.0   
3       5008808           F            N               Y           0.0   
4       5008809           F            N               Y           0.0   
...         ...         ...          ...             ...           ...   
438552  6840104           M            N               Y           0.0   
438553  6840222           F            N               N           0.0   
438554  6841878           F            N               N           0.0   
438555  6842765           F            N               Y           0.0   
438556  6842885           F            N               Y           0.0   

        AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0               427500.0      

There are 438557 rows in df_application and there are 1,048,575 rows in df_credit. 

In [30]:
df_application['ID'].nunique()

438510

In [31]:
df_credit['ID'].nunique()

45985

There are 438510 unique bank clients in df_application and 45985 unique bank clients in df_credit

In [37]:
df = pd.merge(df_application, df_credit, on='ID', how='inner')
df

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0       5008804           M            Y               Y           0.0   
1       5008804           M            Y               Y           0.0   
2       5008804           M            Y               Y           0.0   
3       5008804           M            Y               Y           0.0   
4       5008804           M            Y               Y           0.0   
...         ...         ...          ...             ...           ...   
777710  5150337           M            N               Y           0.0   
777711  5150337           M            N               Y           0.0   
777712  5150337           M            N               Y           0.0   
777713  5150337           M            N               Y           0.0   
777714  5150337           M            N               Y           0.0   

        AMT_INCOME_TOTAL NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0               427500.0          Working               Higher education   
1               427500.0          Working               Higher education   
2               427500.0          Working               Higher education   
3               427500.0          Working               Higher education   
4               427500.0          Working               Higher education   
...                  ...              ...                            ...   
777710          112500.0          Working  Secondary / secondary special   
777711          112500.0          Working  Secondary / secondary special   
777712          112500.0          Working  Secondary / secondary special   
777713          112500.0          Working  Secondary / secondary special   
777714          112500.0          Working  Secondary / secondary special   

          NAME_FAMILY_STATUS NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0             Civil marriage  Rented apartment      -12005          -4542   
1             Civil marriage  Rented apartment      -12005          -4542   
2             Civil marriage  Rented apartment      -12005          -4542   
3             Civil marriage  Rented apartment      -12005          -4542   
4             Civil marriage  Rented apartment      -12005          -4542   
...                      ...               ...         ...            ...   
777710  Single / not married  Rented apartment       -9188          -1193   
777711  Single / not married  Rented apartment       -9188          -1193   
777712  Single / not married  Rented apartment       -9188          -1193   
777713  Single / not married  Rented apartment       -9188          -1193   
777714  Single / not married  Rented apartment       -9188          -1193   

        FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0                1                1           0           0             NaN   
1                1                1           0           0             NaN   
2                1                1           0           0             NaN   
3                1                1           0           0             NaN   
4                1                1           0           0             NaN   
...            ...              ...         ...         ...             ...   
777710           1                0           0           0        Laborers   
777711           1                0           0           0        Laborers   
777712           1                0           0           0        Laborers   
777713           1                0           0           0        Laborers   
777714           1                0           0           0        Laborers   

        CNT_FAM_MEMBERS  MONTHS_BALANCE STATUS  
0                     2               0      C  
1                     2              -1      C  
2                     2              -2      C  
3                     2              -3      C  
4                     2              -4      C  
...                 ...             ...    

In [35]:
df['ID'].nunique()

36457

There are 777715 rows and 36457 unique 

---

**Question 2.**

1. Change the values of `STATUS` in `df` according to the following mapping: {C, X, 0} -> 0 and {1, 2, 3, 4, 5} -> 1 making sure that the new values of 0 and 1 are encoded as integers. (2 marks)
2. Create a new numpy array named `list_of_defaults` containing *unique* `ID` numbers for the clients who have `STATUS` = 1 in any of the last 12 months in the dataset. (2 marks) 
3. Create a new DataFrame called `df_final` that contains the rows of `df` for which the `ID` are in `list_of_defaults`, keeping only one row for each `ID` (i.e. eliminate rows with duplicate `ID`s while keeping the first duplicate row). How many rows do you have in `df_final`? Answer using both print() function and in Markdown text. (Hint: find out about `isin()` function in pandas.) (2 marks)
4. Add a new column `y = 1` for all the rows in `df_final`. (1 marks)
5. Increase `df_final` to a total of 4,000 rows by adding rows from `df` with unique `ID`s (nonduplicated `ID`s) which are not in `list_of_defaults`. To do this start adding the rows from the beginning of `df`. (Hint: learn what `~`, i.e. tilde sign, does in pandas). (2 marks) 
6. Fill the missing values of `y` in `df_final` with zeros. Remove `STATUS` and `MONTHS_BALANCE` from `df_final`. How many clients with  overdue payments of more than 29 days and how many clients with less than 29 days overdue payments are there in `df_final`? Answer using both print() function and in Markdown text.(1 mark)

(10 marks)

In [40]:
status_mapping = {'C': 0, 'X': 0, '0': 0, '1': 1, '2': 1, '3': 1, '4': 1, '5': 1} #Qu1

df['STATUS'] = df['STATUS'].replace(status_mapping).astype(int)

In [42]:
df

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0       5008804           M            Y               Y           0.0   
1       5008804           M            Y               Y           0.0   
2       5008804           M            Y               Y           0.0   
3       5008804           M            Y               Y           0.0   
4       5008804           M            Y               Y           0.0   
...         ...         ...          ...             ...           ...   
777710  5150337           M            N               Y           0.0   
777711  5150337           M            N               Y           0.0   
777712  5150337           M            N               Y           0.0   
777713  5150337           M            N               Y           0.0   
777714  5150337           M            N               Y           0.0   

        AMT_INCOME_TOTAL NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0               427500.0          Working               Higher education   
1               427500.0          Working               Higher education   
2               427500.0          Working               Higher education   
3               427500.0          Working               Higher education   
4               427500.0          Working               Higher education   
...                  ...              ...                            ...   
777710          112500.0          Working  Secondary / secondary special   
777711          112500.0          Working  Secondary / secondary special   
777712          112500.0          Working  Secondary / secondary special   
777713          112500.0          Working  Secondary / secondary special   
777714          112500.0          Working  Secondary / secondary special   

          NAME_FAMILY_STATUS NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0             Civil marriage  Rented apartment      -12005          -4542   
1             Civil marriage  Rented apartment      -12005          -4542   
2             Civil marriage  Rented apartment      -12005          -4542   
3             Civil marriage  Rented apartment      -12005          -4542   
4             Civil marriage  Rented apartment      -12005          -4542   
...                      ...               ...         ...            ...   
777710  Single / not married  Rented apartment       -9188          -1193   
777711  Single / not married  Rented apartment       -9188          -1193   
777712  Single / not married  Rented apartment       -9188          -1193   
777713  Single / not married  Rented apartment       -9188          -1193   
777714  Single / not married  Rented apartment       -9188          -1193   

        FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0                1                1           0           0             NaN   
1                1                1           0           0             NaN   
2                1                1           0           0             NaN   
3                1                1           0           0             NaN   
4                1                1           0           0             NaN   
...            ...              ...         ...         ...             ...   
777710           1                0           0           0        Laborers   
777711           1                0           0           0        Laborers   
777712           1                0           0           0        Laborers   
777713           1                0           0           0        Laborers   
777714           1                0           0           0        Laborers   

        CNT_FAM_MEMBERS  MONTHS_BALANCE  STATUS  
0                     2               0       0  
1                     2              -1       0  
2                     2              -2       0  
3                     2              -3       0  
4                     2              -4       0  
...                 ...             .

In [43]:
df_filtered = df[(df['STATUS'] == 1) & (df['MONTHS_BALANCE'] >= -12) & (df['MONTHS_BALANCE'] <= 0)] #Qu2

In [45]:
list_of_defaults = np.unique(df_filtered['ID'].values)

In [52]:
df_final = df[df['ID'].isin(list_of_defaults)].drop_duplicates(subset='ID', keep='first') #3

In [54]:
print(len(df_final))

1833


Number of rows in df_final: 1833

In [64]:
df_final['y'] = 1 #Qu4

In [65]:
filtered_df = df[~df['ID'].isin(list_of_defaults)].groupby('ID').first().reset_index() #Qu5

In [66]:
df_final = pd.concat([df_final, filtered_df], ignore_index=True)

In [67]:
df_final['y'].fillna(0, inplace=True) #Qu6

In [68]:
df_final.drop(columns=['STATUS', 'MONTHS_BALANCE'], inplace=True)

In [75]:
overdue_more = df_final[df_final['y'] > 29]['ID'].nunique()
print(overdue_more)
overdue_less = df_final[df_final['y'] <= 29]['ID'].nunique()
print(overdue_less)

0
36457


There are 0 people with overdue payments of more than 29 days.
There are 36457 people with overdue payments of less than 29 days

<hr style="width:25%;margin-left:0;"> 

**Question 3**. 
- Delete `ID` column from `df_final` (1 marks)
- Of the remaining variables in `df_final` and assuming that `NAME_EDUCATION_TYPE` is the only ordinal variable, how many variable are of numeric and nominal types? Provide lists of all numeric and nominal variables. (6)
- Using an appropriate function find and comment on the missing values in `df_final`, i.e. how many variables and how many observations? (3 marks)   
(10 marks)

In [76]:
df_final.drop(columns=['ID'], inplace=True)

In [77]:
numeric_vars = df_final.select_dtypes(include=[np.number]).columns.tolist()
nominal_vars = df_final.select_dtypes(include=[object]).columns.tolist()

The number of numeric variables are 10 and the number of nominal variables are 8.

The list of numeric variables are: 'CNT_CHILDREN', 'AMT_INCOME_TOTAL', 'DAYS_BIRTH', 'DAYS_EMPLOYED', 'FLAG_MOBIL', 'FLAG_WORK_PHONE', 'FLAG_PHONE', 'FLAG_EMAIL', 'CNT_FAM_MEMBERS', 'y'

The list of nominal variables are: 'CODE_GENDER', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'NAME_INCOME_TYPE', 'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 'OCCUPATION_TYPE'

In [82]:
missing_values = df_final.isna().sum()
num_observations, num_variables = df_final.shape

There are a total of 221 missing values in 'CNT_CHILDREN', 5331 missing values in 'NAME_EDUCATION_TYPE', 32957 missing values in 'OCCUPATION_TYPE' and no missing values in the rest of the columns. 

There are 105705 observations and 18 variables

---
---

### Task 2: Imputing missing values and dealing with categorical features (Total Marks: 30)



**Question 4.** 
- Use an appropriate `pandas` function to impute missing values in `df_final` (10 marks)
    - Be careful when deciding which method to use to replace missing observations 
    - Take into consideration the type of each variable and the best practices we discussed in class/lecture notes
- Briefly explain what you have done and why. (5 marks)

(Total: 15 marks)

In [88]:
df_final['CNT_CHILDREN'].fillna(0, inplace=True)
df_final['NAME_EDUCATION_TYPE'].fillna(df_final['NAME_EDUCATION_TYPE'].mode()[0], inplace=True)
df_final['OCCUPATION_TYPE'].fillna(df_final['OCCUPATION_TYPE'].mode()[0], inplace=True)

For the variable 'CNT_CHILDREN', I impute the missing values with the value 0, assuming these clients do not have children. 

For the variables 'NAME_EDUCATION_TYPE' and 'OCCUPATION_TYPE' 



For the target variable 'y', we chose to impute with zeros, assuming missing values mean no overdue payment (binary variable).
For the numeric variable 'AGE', we imputed missing values with the mean value of the available data. This is a common practice for continuous variables.
For the numeric variable 'INCOME', we imputed missing values with the median value of the available data. Median is less sensitive to outliers.
For the categorical variables 'NAME_EDUCATION_TYPE' and 'GENDER', we imputed missing values with the mode (most frequent value) of the respective variables.
Imputing based on variable types and best practices helps maintain the integrity of the data and ensures that imputed values are reasonable representations of the missing data.

<hr style="width:25%;margin-left:0;"> 

**Question 5**. Convert the values in `NAME_EDUCATION_TYPE` as follows
- Lower secondary -> 1
- Secondary / secondary special -> 2
- Incomplete higher -> 3
- Higher education -> 4


(Total: 5 marks)  

In [90]:
education_type_mapping = {
    'Lower secondary': 1,
    'Secondary / secondary special': 2,
    'Incomplete higher': 3,
    'Higher education': 4}

In [93]:
df_final['NAME_EDUCATION_TYPE'] = df_final['NAME_EDUCATION_TYPE'].replace(education_type_mapping)

<hr style="width:25%;margin-left:0;"> 

**Question 6**. 

Add dummy variables to `df_final` for all of the nominal features which are currently stored as string (text). 
- Make sure to delete the original variables from the dataframe
- Drop the first column from each set of created dummy variable, i.e. for each feature



(Total: 10 marks)  

In [95]:
df_final = pd.get_dummies(df_final, columns=['CODE_GENDER', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'NAME_INCOME_TYPE', 'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 'OCCUPATION_TYPE'], drop_first=True)


print(df_final)

KeyError: "None of [Index(['CODE_GENDER', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'NAME_INCOME_TYPE',\n       'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 'OCCUPATION_TYPE'],\n      dtype='object')] are in the [columns]"

---
---

### Task 3 Preparing X and y arrays (Total Marks: 10)

**Question 7**. 

- Create a numpy array named `y` from the `y` column of `df_final` making sure that the values of the array `y` are stored as integers (3 marks)   
- Create a numpy array named `X`  from all the remaining features in `df_final` (2 marks)   

(Total: 5 Marks)

In [20]:
# ---- provide your code here -----

<hr style="width:25%;margin-left:0;"> 

**Question 8**. 

- Use an appropriate scikit-learn library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% train and 30% test datasets (2.5 marks) 
    - Set random_state to 7 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
- Standardise the data using `StandardScaler` library (2.5 marks)   

(Total: 5 marks) 

In [21]:
# ---- provide your code here -----

---
---

### Task 4. Support Vector Classifier and Accuracies (Total Marks: 30)


**Question 9**. 

- Train a Support Vector Classifier on standardised data (3 marks)
    - Use `linear` kernel and set `random_state` to 7 (don't change any other parameters)
    - Compute and print training and test dataset accuracies
- Train another Support Vector Classifier on standardised data (3 marks)
    - Use `rbf` kernel and set `random_state` to 7 (don't change any other parameters)
    - Compute and print training and test dataset accuracies
- What can you say about the presence of nonlinearities in the dataset? (4 marks)

(Total: 10 marks)  

In [22]:
# ---- provide your code here -----

---- provide your text answer here ----

<hr style="width:25%;margin-left:0;"> 

**Question 10**

- Extract 2 linear principal components from the standardised features using an appropriate `sklearn` library (5 marks)
- Train a Support Vector Classifier on the computed principal components (5 marks) 
    - Use `rbf` kernel and set `random_state` to 7 (don't change any other parameters)
- Compute and print training and test dataset accuracies (5 marks)
- What can you say about the ability of the 2 principal components to compress the information contained in the features matrix `X`, and why? (5 marks)     


(Total: 20 marks)  

In [23]:
# ---- provide answer here -----

---- provide your text answer here ----


---
---